src/python_testing/TC_DeviceConformance.py

```diff
@@ -373,6 +414,36 @@ def record_warning(location, problem):
 
         return success, problems
 
+    def check_root_endpoint_for_application_device_types(self) -> list[ProblemNotice]:
+        problems = []
+        device_types = [d.deviceType for d in self.endpoints[0][Clusters.Descriptor][Clusters.Descriptor.Attributes.DeviceTypeList]]
+
+        for d in device_types:
+            if self.xml_device_types[d].classification_class.lower() == 'simple':
+                location = DeviceTypePathLocation(device_type_id=d)
+                problems.append(ProblemNotice("TC-DESC-2.3", location, ProblemSeverity.ERROR,
+                                f"Application device type {self.xml_device_types[d].name} found on EP0"))
+
+        return problems
+
+    def check_all_application_device_types_superset(self) -> list[ProblemNotice]:
+        problems = []
+        supersets = get_supersets(self.xml_device_types)
+        for endpoint_num, endpoint in self.endpoints.items():
+            all_device_type_ids = [dt.deviceType for dt in endpoint[Clusters.Descriptor]
+                                   [Clusters.Descriptor.Attributes.DeviceTypeList]]
+            application_device_type_ids = set([
+                dt for dt in all_device_type_ids if self.xml_device_types[dt].classification_class == 'simple'])
+            if len(application_device_type_ids) <= 1:
+                continue
+            if any([application_device_type_ids.issubset(superset) for superset in supersets]):
+                continue
+
+            location = AttributePathLocation(3, Clusters.Descriptor.id, Clusters.Descriptor.Attributes.DeviceTypeList.attribute_id)
+            problems.append(ProblemNotice('TC-DESC-2.3', location=location, severity=ProblemSeverity.ERROR,
```

Comment:
would it be possible to log what we think are the conflicts?

Maybe we would need `[s for s in supersets if aplication_device_type_ids & s]` to show the things that intersect with our sets.

Reply:
discussed offline - will miss the non-superset device types. Folks will just have to figure it out - the device types are listed in the problem notice.